routes/index.js

```diff
@@ -4,10 +4,11 @@ var Product = require('../models/product');
 
 /* GET home page. */
 router.get('/', function(req, res, next) {
-  res.render('index', { products: /* ??? */ });
+  res.render('index', { products: Product.all});
 });
 
 router.post('/cart', function(req, res, next) {
+var cookie = product.id;
```

Comment:
So I think in this section, we might benefit from thinking through *exactly* what it is we want the outcome of this route to be before we start to code it. What should the `post` route to `/cart` do? Well we know that `post` routes are for creating data, and the thing we're creating or sending to is the `/cart` path, so I think we can reasonably assume that the goal of this route is to add an item to the cart.

So then how do we do that? What things do we need in order to even begin to add an item to a cart? Well, I think at the very least we need an item, and we need a a cart to add it to. Let's start with the item. How do we know what item it is? Do we have any identifying information being sent to us? If we check the parameters sent to us by the user, we see that `req.body` should contain an ID number. If we see where that ID comes from, it's from one of the forms on the `index.ejs` page to add a cookie to the cart. So presumably the ID number is the ID of a cookie that the user wants to add to their cart. So once we have that ID, we can hopefully use it to find the cookie we want - if you look in the `product.js` file, you'll see there's a method just for this purpose called `Product.find`. If we pass `Product.find` an ID number, it will give us back the cookie that corresponds to that ID (if it exists). So if we do something like

```js
var cookie = Product.find(parseInt(req.body.id));
```

then the `cookie` variable should contain a cookie object (we need to `parseInt` the ID because it came to us as a string, but we need it to be a number).

So we have the item (the cookie) we want to add to add to the cart - but what about the cart itself? This is where sessions come in. We need some sort of data structure to hold a list of all the items in our cart, and we need that list to be persisted across page refreshes. So in our `app.js` file we have already required and initialized the code for using sessions - now we just need to use them. Similar to how we did it in previous examples with sessions, we probably want something like the following in our `app.js` file:

```js
app.use(function(req, res, next) {
  req.session.cart = req.session.cart || [];
  app.locals.cart = req.session.cart;
});
```

This bit of code does 2 things. First and most important, it sets a property on the `req.session` object called `cart` that we can now use in any of our routes. The `cart` property gets set to itself if it already exists, or an empty array if it doesn't exist - this is so if we already created a cart, we don't overwrite it, but if the user doesn't have a cart we initialize the cart to an empty array. Either way, at the end of the line we have a `cart` property we can access that is an array. The next line of code adds a property to the `locals` object that makes it available to our views - because the `req` object isn't available in our views, we need some way to reference the items in the cart without using `req`, and that's what `app.locals.cart` is for.

Finally, once we've created our `cart` property in `app.js`, we can come back to our route over here and use it. Now that we have our cookie stored in a `cookie` variable, we can do something like `req.session.cart.push(cookie)` to add the cookie to the cart. So the whole route looks like:

```js
router.post('/cart', function(req, res, next) {
  var cookie = Product.find(parseInt(req.body.id));
  req.session.cart.push(cookie);
  res.redirect('/');
});
```

So we find the cookie, we add it to the cart, then we redirect back to our main `GET` route to display the list of available products again, along with the items in the cart (which it looks like you started to do in the `index.ejs` file).